engineering-plans/0003-remove-jsonb-storage.md

```diff
@@ -0,0 +1,125 @@
+and can be done in parallel.
+
+## Implementation
+
+### User Subgraphs in the Hosted Service
+
+We will need to communicate to users that they need to update their
+subgraphs if they still use JSONB storage. Currently, there are ~ 580
+subgraphs
+([list](https://gist.github.com/lutter/2e7a7716b70b4144fe0b6a5f1c9066bc))
+belonging to 220 different organizations using JSONB storage. It is quite
+likely that the vast majority of them is not needed anymore and simply left
+over from somebody trying something out.
+
+We should contact users and tell them that we will delete their subgraph
+after a certain date (say 2020-02-01) _unless_ they deploy a new version of
+the subgraph (with an explanation why etc. of course) Redeploying their
+subgraph is all that is needed for those updates.
+
+### Self-hosted User Subgraphs
+
+We will need to tell users that the 'old' JSONB storage is deprecated and
+support for it will be removed as of some target date, and that they need
+to redeploy their subgraph.
+
+Users will need some documentation/tooling to help them understand
+* which of their deployed subgraphs still use JSONB storage
+* how to remove old subgraphs
+* how to remove old deployments
+
+### Subgraph Metadata in `subgraphs.entities`
+
+We can treat the `subgraphs` schema like a normal subgraph, with the
+exception that some entities must not be versioned. For that, we will need
+to adopt code that makes it possible to write entities to the store without
+recording their version (or, more generally, so that there will only be one
+version of the entity, tagged with a block range `[0,)`)
+
+We will manually create the DDL for the `subgraphs.graphql` schema and run
+that as part of a database migration. In that migration, we will also copy
+the existing metadata from `subgraphs.entities` and
+`subgraphs.entity_history` into their new tables.
+
+### The Code Base
+
+Delete all code handling JSONB storage. This will mostly affect
+`entities.rs` and `jsonb_queries.rs` in `graph-store-postgres`, but there
+are also smaller things like that we do not need the annotations on
+`Entity` to serialize them to the JSON format that JSONB uses.
+
+## Tests
+
+Most of the code-level changes are covered by the existing test suite. The
+major exception is that the migration of subgraph metadata needs to be
+tested and checked manually, using a recent dump of the production
+database.
+
+## Migration
+
+See above on migrating data in the `subgraphs` schema.
+
+## Documentation
+
+No user-facing documentation is needed.
+
+## Implementation Plan
+
+_No estimates yet as we should first agree on this general course of
+action_
+
+* Notify hosted users to update their subgraph or have it deleted by date X
+* Mark JSONB storage as deprecated and announce when it will be removed
+* Provide tool to ship with `graph-node` to delete unused deployments and
+  unneeded subgraphs
+* Add affordance to not version entities to relational storage code
```

Comment:
If we make this per-entity, then we could also use it for the network indexer.

Reply:
Yes, it will have to be per entity, since we have metadata that is unversioned (belonging to a static data source) and versioned (belonging to a dynamic data source), e.g., `EthereumContractSource`